contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,348 @@
+ 
+ However, by some extension to this effort, we would like to improve scalability
+of the system, by significantly reducing amount of api calls coming from kubelets.
+As of now, to avoid situation that kubelet is watching all secrets/configmaps/...
+in the system, it is not using watch. Instead of that, it is retrieving indidual
+objects, by sending individual GET requests. However, it is sending those requests
+periodically to enable automatic updates of mounted secrets/configmaps/...
+In large clusters, this is generating huge unnecessary load, as this load in
+principle should be watch-based. We would like to address this together with
+solving the authorization issue.
+
+
+# Proposal
+
+ In this proposal, we're not focusing on how exactly security should be done.
+We're just sketching very high level approach and exact authorization mechanism
+should be discussed separately.
+
+ At the high level, what we would like to achieve is to enable LIST and WATCH
+requests to support more sophisticated filtering (e.g. we would like to be able
+to somehow ask for all Secrets attached to all pods bound to a given node in a
+single LIST or WATCH request). We obviously want to be able to authorize all
+other types of requests (in particular GETs), so that the design has to be
+consistent in all those cases.
+
+ To solve this problem, we propose to introduce the idea of `bulk watch ` and
+`bulk get ` (and in the future other bulk operations). This idea was already
+appearing in the community and now it seems to be a good time to proceed with it.
+
+ To allow Kubelet to use watch, we also need to periodically verify ACLs of all
+watches. Whenever a user loses access to a given resource, the watch should be
+close within some bounded time. The rationale behind this requirement is that
+by using (bulk) watch we can't regress the current behavior where a polling
+Kubelet would eventually be unable to access a secret.
+
+TODO: I think this isn't specific to bulk watch and we need to solve this
+problem also in regular watch. For example watcher watching just a single
+secret may also lose access to it and this watch should be closed in this case.
+
+ We would like the API to be generic enough so that it will be useful for many
+different usecases, not just for this particular one. As a result, below we are
+describing requirements that the solution has to meet to satisfy our needs:
+- a single bulk request has to support multiple resources (e.g. get a node and
+all pods associated with it)
+- if the bulk request filters asks only for object of a single resource type,
+the returned result should ideally be comparible with what is returned from
+the regular API
+- the API has to support aggregation if different resource types are server by
+different apiservers
+- ACLs of watches have to be periodically verified and automatically closed if
+ACLs no longer allow for this watch.
+- at some point in the future we want clients to be able to alter their watch
```

Comment:
if a node has to restart all secret watches any time it starts watching a new secret, that is worse that what it is doing today

scenario:
* a new pod referencing a new secret is scheduled to a node every ten seconds
* volume sync occurs every minute
* secret cache TTL is 1 minute

currently:
```
t0:00: get secret 
t0:10: get secret 
t0:20: get secret
t0:30: get secret
t0:40: get secret
t0:50: get secret
t1:00: volume sync refreshes 1 secret (1 gets)
t1:00: get secret 
t1:10: get secret 
t1:20: get secret
t1:30: get secret
t1:40: get secret
t1:50: get secret
t2:00: volume sync refreshes 7 secrets (7 gets)
```

in 2 minutes, we make 20 get requests

if every secret addition dropped the watch and reestablished bulk watches, it would look like this:
```
t0:00: establish watch of 1 secret 
t0:10: establish watch of 2 secrets
t0:20: establish watch of 3 secrets
t0:30: establish watch of 4 secrets
t0:40: establish watch of 5 secrets
t0:50: establish watch of 6 secrets
t1:00: establish watch of 7 secrets
t1:10: establish watch of 8 secrets
t1:20: establish watch of 9 secrets
t1:30: establish watch of 10 secrets
t1:40: establish watch of 11 secrets
t1:50: establish watch of 12 secrets
```

in 2 minutes, we're establishing watches for 78 secrets. even though there are only 12 API connections made, there are 78 authorization checks (that's assuming we can start a watch without doing a get or list first, which I'm unsure of)